csharp/src/Ice/ConnectionI.cs

```diff
@@ -2024,48 +2037,22 @@ private int sendMessage(OutgoingMessage message)
         {
             cancelInactivityTimer();
         }
-        // If we're sending a heartbeat, there is a chance the connection is inactive and that we need to schedule the
-        // inactivity timer. It's ok to do this before actually sending the heartbeat since the heartbeat does not count
-        // as an "activity".
-        else if (
-            _inactivityTimer is null &&           // timer not already scheduled
-            _inactivityTimeout > TimeSpan.Zero && // inactivity timeout is enabled
-            _state == StateActive &&              // only schedule the timer if the connection is active
-            _dispatchCount == 0 &&                // no pending dispatch
-            _asyncRequests.Count == 0 &&          // no pending invocation
-            _readHeader)                          // we're not waiting for the remainder of an incoming message
-        {
-            bool isInactive = true;
 
-            // We may become inactive while the peer is back-pressuring us. In this case, we only schedule the
-            // inactivity timer if all outgoing messages in _sendStreams are heartbeats.
-            foreach (OutgoingMessage queuedMessage in _sendStreams)
+        if (_sendStreams.Count > 0)
+        {
+            if (isHeartbeat)
             {
-                // TODO: temporary work-around for #2336
-                Ice.Internal.Buffer buffer = queuedMessage.stream.getBuffer();
-                if (!buffer.empty()) // should never happen
-                {
-                    if (buffer.b.get(8) != Protocol.validateConnectionMsg)
-                    {
-                        isInactive = false;
-                        break; // for
-                    }
-                }
+                // The message already queued does the job of the heartbeat.
```

Comment:
Interestingly, the heartbeat can still be queued (see end of method). Should we prevent this queuing too?

I think we should as it could be a "false positive": _sendStreams is not empty even though it only contains a heartbeat.

Reply:
Actually, we should not skip a heartbeat when that's the only pending message.

Reply:
The situation where _sendStreams contains only a heartbeat for a while is indicative of an unhealthy connection, which will be soon aborted by the idle check.

Reply:
I am wondering if this skipping is correct:
 - we send a heartbeat to keep the connection alive
 - we find a pending message in _sendStreams so we drop it
 - the application cancels the corresponding invocation so this pending message is never sent
 - no byte gets written so the peer aborts the connection

Reply:
You are right if the queued invocation is about to be canceled and we don't queue the heartbeat, the peer might close the connection.

Reply:
What should we do then?

Iterate over _sendStreams to see if there is already a non-cancellable message? The bulk of this simplification was to avoid this iteration. And checking message types would increase complexity.

cancellation messages = Request, RequestBatch?
non-cancellable message = Reply, ValidateConnection, CloseConnection?

Reply:
Keep in mind that `_sendStreams.First` is the message _being_ sent not the message that will be sent next. The start of its buffer has already been written to the transport but it couldn't be fully sent and the remainder of the message will be sent by `message`.

> The situation where _sendStreams contains only a heartbeat for a while is indicative of an unhealthy connection, which will be soon aborted by the idle check.

Perhaps the transport wasn't ready to send the heartbeat (back pressure from the peer for example). It doesn't imply the connection is unhealthy. It just indicates that the sending of the heartbeat is pending. The chances of this happening with TCP are close to zero but what about other transports?

> I am wondering if this skipping is correct:
> - we send a heartbeat to keep the connection alive
> - we find a pending message in _sendStreams so we drop it
> - the application cancels the corresponding invocation so this pending message is never sent
> - no byte gets written so the peer aborts the connection

I think it's correct because the `_sendStreams.First` message is _being_ sent and its sending can't be canceled by the cancellation of the associated request.

Reply:
> I think it's correct because the _sendStreams.First message is being sent and its sending can't be canceled by the cancellation of the associated request.

According to https://github.com/zeroc-ice/ice/blob/7c1ace30ad1cb5f39b11e9d549aaccfb3fe94e8d/csharp/src/Ice/ConnectionI.cs#L1199

_sendStreams being not empty doesn't guarantee some bytes _will_ be sent. We can be in the situation where everything has been sent and _sendStreams.First has not been removed yet.

Reply:
> _sendStreams being not empty doesn't guarantee some bytes will be sent. We can be in the situation where everything has been sent and _sendStreams.First has not been removed yet.

Why does it matter here?